hledger-lib/Hledger/Data/Types.hs

```diff
@@ -733,18 +735,31 @@ nullaccountdeclarationinfo = AccountDeclarationInfo {
 
 -- | An account, with its balances, parent/subaccount relationships, etc.
 -- Only the name is required; the other fields are added when needed.
-data Account = Account {
-   aname                     :: AccountName    -- ^ this account's full name
+data Account a = Account {
+   aname                     :: AccountName         -- ^ this account's full name
   ,adeclarationinfo          :: Maybe AccountDeclarationInfo  -- ^ optional extra info from account directives
   -- relationships in the tree
-  ,asubs                     :: [Account]      -- ^ this account's sub-accounts
-  ,aparent                   :: Maybe Account  -- ^ parent account
-  ,aboring                   :: Bool           -- ^ used in the accounts report to label elidable parents
+  ,asubs                     :: [Account a]        -- ^ this account's sub-accounts
+  ,aparent                   :: Maybe (Account a)  -- ^ parent account
+  ,aboring                   :: Bool                -- ^ used in the accounts report to label elidable parents
   -- balance information
-  ,anumpostings              :: Int            -- ^ the number of postings to this account
-  ,aebalance                 :: MixedAmount    -- ^ this account's balance, excluding subaccounts
-  ,aibalance                 :: MixedAmount    -- ^ this account's balance, including subaccounts
-  } deriving (Generic)
+  ,abalances                 :: AccountBalances a   -- ^ historical and date-associated account balances
+  } deriving (Generic, Functor)
+
+-- | A component of an 'Account' containing historical 'AccountBalance', and a
+-- map of start dates to 'AccountBalance'.
+data AccountBalances a = AccountBalances {
+   abhistorical :: a            -- ^ historical balance information
+  ,abdatemap    :: IM.IntMap a  -- ^ balance information associated to a start day
+  } deriving (Eq, Functor, Generic)
```

Comment:
What's the reason for using a type parameter ? Do we truly need it ?

How does the "start day" map, with Int keys I assume, work here ?

Reply:
> What's the reason for using a type parameter ? Do we truly need it ?

For budget report, I guess.

Reply:
Yes, the main use is for the budget report. But I think it also simplifies things a bit by exposing the functor, foldable, and traversable interfaces for `AccountBalances`, saving having to write out a lot of boilerplate to perform tasks that re-implement that functionality in a monomorphic container.